File: oop/03-polymorphism/typescript/savingsAccount.ts
/*
 * Demonstrates: Polymorphism - Override
 * Concept: SavingsAccount overrides processWithdrawal to enforce minimum balance
 */

import { Account, TransactionResult } from "./account";

export class SavingsAccount extends Account {
    public readonly minimumBalance: number;

    constructor(accountNumber: string, accountHolder: string, openingBalance: number, minimumBalance: number) {
        super(accountNumber, accountHolder, openingBalance);
        this.minimumBalance = minimumBalance;
    }

    override processWithdrawal(amount: number): TransactionResult {
        if (amount <= 0) {
            return TransactionResult.failed("Amount must be greater than zero.");
        }

        const resultingBalance = this._balance - amount;

        if (resultingBalance < this.minimumBalance) {
            return TransactionResult.failed(`Cannot withdraw. Minimum balance of R${this.minimumBalance.toFixed(2)} required.`);
        }

        this._balance = resultingBalance;
        return TransactionResult.succeeded(`Withdrew R${amount.toFixed(2)}. New balance: R${this._balance.toFixed(2)}`);
    }

    override getAccountType(): string {
        return "Savings Account";
    }
}

File: oop/03-polymorphism/csharp/SavingsAccount.cs
/*
 * Demonstrates: Polymorphism - Override
 * Concept: SavingsAccount overrides ProcessWithdrawal to enforce minimum balance
 */

namespace FinanceDomain.Polymorphism;

public class SavingsAccount : Account
{
    public decimal MinimumBalance { get; }

    public SavingsAccount(string accountNumber, string accountHolder, decimal openingBalance, decimal minimumBalance)
        : base(accountNumber, accountHolder, openingBalance)
    {
        MinimumBalance = minimumBalance;
    }

    public override TransactionResult ProcessWithdrawal(decimal amount)
    {
        if (amount <= 0)
            return TransactionResult.Failed("Amount must be greater than zero.");

        decimal resultingBalance = _balance - amount;

        if (resultingBalance < MinimumBalance)
            return TransactionResult.Failed($"Cannot withdraw. Minimum balance of R{MinimumBalance:F2} required.");

        _balance = resultingBalance;
        return TransactionResult.Succeeded($"Withdrew R{amount:F2}. New balance: R{_balance:F2}");
    }

    public override string GetAccountType() => "Savings Account";
}
